pjmedia/src/pjmedia/sdp_neg.c

```diff
@@ -1793,3 +1929,128 @@ PJ_DEF(pj_status_t) pjmedia_sdp_neg_fmt_match(pj_pool_t *pool,
+    unsigned max_pt_cnt = PJ_ARRAY_SIZE(used_pt->med_pt_map);
+    const pj_str_t telephone_event = pj_str("telephone-event");
+
+    if (pj_strncmp(codec, &telephone_event, telephone_event.slen) == 0)
+        pt_check = PJMEDIA_RTP_PT_TELEPHONE_EVENTS;
+    else
+        pt_check = START_DYNAMIC_PT;
+    
+    for (; i < used_pt_num && i < max_pt_cnt; ++i) {
+        if (pj_strcmp(codec, &used_pt->med_pt_map[i].codec) == 0) {
+            return used_pt->med_pt_map[i].pt;
+        }
+    }
+    i = 0;
+
+    while ((i < used_pt_num) && (i < max_pt_cnt) && (pt_check <= 127))
+    {
+        if (pt_check == used_pt->med_pt_map[i].pt) {
+            pt_check++;
+            i = 0;
+        } else {
+            i++;
+        }
+    }
+    if (pt_check > 127 || i == max_pt_cnt) {
+        /* No more available PT */
+        return 0;
+    }
+
+    used_pt->med_pt_map[used_pt->med_pt_map_cnt].pt = pt_check;
+    pj_strdup(pool, &used_pt->med_pt_map[used_pt->med_pt_map_cnt++].codec, codec);
+
+    return pt_check;
+}
+
+PJ_DEF(pj_status_t) pjmedia_sdp_neg_validate_pt(pjmedia_sdp_neg* neg,
+                                                pjmedia_sdp_session* sdp_sess)
+{   
+#if PJMEDIA_SDP_NEG_MAINTAIN_SESSION_PT
+    unsigned i = 0;
+    for (; i < sdp_sess->media_count && i < neg->sess_med_pt_map_cnt; ++i) {
+        unsigned j = 0;
+        pjmedia_sdp_attr* attr_tmp[PJMEDIA_MAX_SDP_ATTR];
+        unsigned attr_tmp_cnt = 0;
+        media_pt_map used_pt_map;
+        media_pt_map* med_pt_map = get_media_pt_map(neg, &sdp_sess->media[i]->desc.media);
+
+        if (!med_pt_map)
+            return PJ_ENOTFOUND;
+
+        pj_memcpy(&used_pt_map, med_pt_map, sizeof(used_pt_map));
+        pjmedia_sdp_media* sess_media = sdp_sess->media[i];
+
+        for (; j < sess_media->desc.fmt_count; ++j) {
+            pjmedia_sdp_attr* attr = NULL;
+            pj_str_t pt_str = sess_media->desc.fmt[j];
+            unsigned pt = pj_strtoul(&pt_str);
+
+            if (pj_strtoul(&sess_media->desc.fmt[j]) < START_DYNAMIC_PT)
+                continue;
+
+            attr = pjmedia_sdp_media_find_attr2(sess_media, "rtpmap", &pt_str);
+            if (attr) {
+                pj_str_t codec;
+                unsigned p_exists;
+                unsigned pt_len = sess_media->desc.fmt[j].slen + 1;
+
+                codec.ptr = attr->value.ptr + pt_len;
+                codec.slen = attr->value.slen - pt_len;
+
+                p_exists = pt_exists(&used_pt_map, pt, &codec);
+                if (p_exists == 1 || p_exists == 2) {
+                    /* pt is assigned to another codec, find existing pt used by the
+                     * codec or a new one.
+                     */
+                    unsigned new_pt;
+                    pj_str_t new_pt_str;
+
+                    new_pt = find_valid_pt(neg->pool_active, &used_pt_map, &codec);
+                    if (new_pt == 0)
+                        return PJ_ENOTFOUND;
```

Comment:
- Print log too? Error code can be confusing in debugging.
- Upon this failure, some codecs may have been rewritten while others leaved as is, is there a chance that a media SDP contains two codecs with the same PT and how it is handled?
- Better add comment/docs here/somewhere about the behavior if there is no more valid PT.

Reply:
when there's a failure in finding a new pt, then the previous changes (media format) will be reverted.